Cleanup code that checks for updates

donatello opened the pull request:
## Description
<!--- Describe your changes in detail -->
This patch introduces a new type `updateParams` to parameterize the
code that checks for updates.

* Eliminate passing variables as-is across multiple function (e.g. the
  `mode` string in the earlier code).

* Remove trivial functions (e.g. earlier version of `DownloadReleaseData()` and
  `IsSourceBuild()`).

* Maintain testability of the code.

## Motivation and Context
<!--- Why is this change required? What problem does it solve? -->
<!--- If it fixes an open issue, please link to the issue here. -->

Cleanup for better readability and maintainability.

## How Has This Been Tested?
<!--- Please describe in detail how you tested your changes. -->
<!--- Include details of your testing environment, and the tests you ran to -->
<!--- see how your change affects other areas of the code, etc. -->
Existing test cases pass (`make test`)


## Types of changes
<!--- What types of changes does your code introduce? Put an `x` in all the boxes that apply: -->
- [ ] Bug fix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)

## Checklist:
<!--- Go over all the following points, and put an `x` in all the boxes that apply. -->
<!--- If you're unsure about any of these, don't hesitate to ask. We're here to help! -->
- [x] My code follows the code style of this project.
- [ ] My change requires a change to the documentation.
- [ ] I have updated the documentation accordingly.
- [ ] I have added tests to cover my changes.
- [x] All new and existing tests passed.

nitisht commented on cmd/update-main.go:
```diff
@@ -68,90 +68,112 @@ const (
+// isDocker - returns if the environment is docker or not.
+func (u *updateParams) isDocker() bool {
+	found, err := u.isDockerInternal()
 	fatalIf(err, "Error in docker check.")
```
Probably a better error message here will be `Error in reading cgroup file` ?

donatello replied:
Improved it - we dont have to make it specific to the cgroup file as the check may change in the future.

krishnasrinivas commented on cmd/update-main.go:
```diff
@@ -68,90 +68,110 @@ const (
+	cgroupFile         string
+	releaseChecksumURL string
+	serverMode         string
+}
```
`updateParams` does not look like the right name for this.
i.e For example why is cgroupFile part of updateParams
why is isDockerInternal a method on updateParams

If the structure name and method names are changed so that they look related then it will look better

donatello replied:
Added more comments - this set of parameters enables writing tests for various pieces of logic. It should be more readable now.

balamurugana commented on cmd/update-main.go:
```diff
@@ -68,90 +68,110 @@ const (
-func getCurrentReleaseTime(minioVersion, minioBinaryPath string) (releaseTime time.Time, err error) {
-	if releaseTime, err = time.Parse(time.RFC3339, minioVersion); err == nil {
+type updateParams struct {
+	versionStr         string
```
`version` is better than `versionStr`.  `version` is anyway `string` type.

donatello replied:
:+1:

balamurugana commented on cmd/update-main.go:
```diff
@@ -68,90 +68,110 @@ const (
+
+// newUpdateParams - returns appropriate default value for
+// updateParams
+func newUpdateParams() *updateParams {
```
Why pointer is returned than value?

donatello replied:
Refactored a bit more and eliminated this function.

balamurugana commented on cmd/update-main.go:
```diff
@@ -68,90 +68,110 @@ const (
+// For any change here should be discussed by openning an issue at
+// https://github.com/minio/minio/issues.
+func (u *updateParams) getUserAgent() string {
+	uaPieces := []string{runtime.GOOS, runtime.GOARCH}
```
`tokens` or `envTokens` is better than `uaPieces`

donatello replied:
:+1: for envTokens

balamurugana commented on cmd/update-main.go:
```diff
@@ -68,90 +68,110 @@ const (
 
-func getCurrentReleaseTime(minioVersion, minioBinaryPath string) (releaseTime time.Time, err error) {
-	if releaseTime, err = time.Parse(time.RFC3339, minioVersion); err == nil {
+type updateParams struct {
```
How about `updateChecker` than `updateParams`?

donatello replied:
Changed to the more explicit `updateCheckParams` :)

balamurugana commented on cmd/update-main.go:
```diff
@@ -68,90 +68,110 @@ const (
-
+// isDocker - returns if the environment is docker or not.
+func (u *updateParams) isDocker() bool {
+	found, err := u.isDockerInternal()
```
`isDockerEnv()` is better than `isDockerInternal()`.

donatello replied:
:+1:

balamurugana commented on cmd/update-main.go:
```diff
@@ -235,20 +257,22 @@ func getDownloadURL() (downloadURL string) {
 
-func getUpdateInfo(timeout time.Duration, mode string) (older time.Duration, downloadURL string, err error) {
-	currentReleaseTime, err := GetCurrentReleaseTime()
+func (u *updateParams) getUpdateInfo(timeout time.Duration) (
```
IMO this method needs to be exported ie `GetUpdateInfo()`.  This will give better readability at consumer side.  but its your call.

donatello replied:
Changed it to not depend on the struct - since the struct can be initialized here. No need to export I think unless we are following a particular convention.

balamurugana reviewed (changes requested):
Minor comments

balamurugana commented on cmd/server-main.go:
```diff
@@ -80,7 +80,7 @@ EXAMPLES:
 func checkUpdate(mode string) {
 	// Its OK to ignore any errors during getUpdateInfo() here.
-	if older, downloadURL, err := getUpdateInfo(1*time.Second, mode); err == nil {
+	if older, downloadURL, err := getUpdateInfo(mode, 1*time.Second); err == nil {
```
Now I don't see any improvement/fixes the PR does than another way to implement with a struct.

donatello replied:
< moved comment to main-thread so as to not lose it when/if the commit is modified >

balamurugana commented on cmd/server-main.go:
```diff
@@ -80,7 +80,7 @@ EXAMPLES:
 func checkUpdate(mode string) {
 	// Its OK to ignore any errors during getUpdateInfo() here.
-	if older, downloadURL, err := getUpdateInfo(1*time.Second, mode); err == nil {
+	if older, downloadURL, err := getUpdateInfo(mode, 1*time.Second); err == nil {
```
As `mode` and `timeout` are required to be passed to `getUpdateInfo()`, No point in changing the order.

donatello replied:
Done.

balamurugana commented on cmd/update-main.go:
```diff
@@ -68,90 +68,108 @@ const (
+// version of the current server process is older than the latest
+// available official release. They enable the testability of various
+// routines in the update checking process.
+type updateCheckParams struct {
```
the mid-path name `updateCheckParams` looks odd to me.  As `getUpdateInfo()` is not a method anymore, `updateParams` is enough.

donatello replied:
No, I think making the name explicit is good - it just needs to be read as `update-check params` meaning "parameters used for doing an update-check". The comment also makes this clear.

balamurugana commented on cmd/update-main.go:
```diff
@@ -68,90 +68,108 @@ const (
+// getCurrentReleaseTime - returns this process's release time.  If it
+// is official minio version, parsed version is returned else minio
+// binary's mod time is returned.
+func (u *updateCheckParams) getCurrentReleaseTime() (releaseTime time.Time,
```
Having these methods to pointer is incorrect.

donatello replied:
Why is it incorrect?

balamurugana replied:
You can try below
```go
package main

import (
	"fmt"
)

type mytype struct {
	name string
}

func (t mytype) hello() {
	fmt.Println("hello", t.name)
}

func (t *mytype) welcome() {
	fmt.Println("welcome", t.name)
}

func main() {
	t1 := mytype{"world"}
	t1.hello()
	t1.welcome()

	var t2 *mytype
	t2.hello()
	t2.welcome()
}
```

donatello replied:
In the above code sample, `t2` is not assigned, so it crashes. Seems like an argument to not have pointer receivers in the Go language! Not sure what your point is, because in my code, the struct is initialized in only place in the application (and few more times in tests), in the same file and not used elsewhere. Obviously it cannot be a nil pointer (by trivial inspection).

Doing some research on this, I found review guidelines in the [Go Wiki](https://github.com/golang/go/wiki/CodeReviewComments#receiver-type) - some cherry-picked points that seem relevant (emphasis my own):

> * If the receiver is a large struct or array, a pointer receiver is more efficient. How large is large? Assume it's equivalent to passing all its elements as arguments to the method. If that feels too large, it's also too large for the receiver.
> * If the receiver is a small array or struct that is naturally a value type (for instance, something like the time.Time type), with no mutable fields and no pointers, or is just a simple basic type such as int or string, a value receiver makes sense. A value receiver can reduce the amount of garbage that can be generated; if a value is passed to a value method, an on-stack copy can be used instead of allocating on the heap. (The compiler tries to be smart about avoiding this allocation, but it can't always succeed.) **Don't choose a value receiver type for this reason without profiling first.**
> * Finally, when in doubt, use a pointer receiver.
> 

Though this is not performance sensitive, for curiosity sake, I did some profiling too using this code https://play.golang.org/p/PAuPwIWojW - it does not run in play.golang.org (takes too long), but running on my laptop I get:

```
$ ./testprog 
PointerReceiverResult
2000000000	         0.41 ns/op
ValueReceiverResult
500000000	         3.05 ns/op
```

The program uses a struct that is identical to `updateCheckParams` (has 5 string properties). Clearly the pointer receiver method is faster by like 7X. In summary, it is ok to use pointer receivers here.

balamurugana replied:
* As you told WRT `updateCheckParams` is not performance sensitive. the pointer type is good for methods changing fields of the type.  This case is not applicable to the struct as none of the methods are updating fields.

* As I indicated in the example, pointer type must be initialized to avoid run time panic.  This is avoidable if methods are to value type.

* method to pointer type is even accessible by `nil` pointer.   Having any methods to pointer type will lead to panic.
```go
func (t *mytype) greet() {
	fmt.Println("spock hand")
}

var t2 *mytype
t1.greet()
```

* Choosing pointer or value type for method is by considering performance and/or changing field values.

donatello replied:
> * As you told WRT updateCheckParams is not performance sensitive. the pointer type is good for methods changing fields of the type. This case is not applicable to the struct as none of the methods are updating fields.

Pointer receivers are **required** for methods changing fields of the struct/type. There is no choice in the matter there. The choice is only in the case of methods not updating the struct/type they are attached to.

So what are the considerations in choosing between pointer/value receiver types for such methods? The [Go Code Review Comments](https://github.com/golang/go/wiki/CodeReviewComments#receiver-type) say that it makes sense for small structs. It also says that you have to think of it as passing all properties of the struct as arguments to each value-receiver method. To me passing 5 variables to so many methods is excessive. So I find a pointer type to be better here. It is not a performance sensitivity argument - but more of an aesthetic to not be wasteful where it is trivial to do so.

You say having pointer receiver has the issue of passing an uninitialized pointer to a method:

> * As I indicated in the example, pointer type must be initialized to avoid run time panic. This is avoidable if methods are to value type.

However, such programmer mistakes are not easily overcome by simply robotically following a convention - see this code:

``` go
package main

import (
	"fmt"
)

type mytype struct{}

func (t mytype) greet() {
	fmt.Println("spock hand")
}

func doSomething(k mytype) {
	k.greet()
}

func main() {
	var t1 *mytype
	doSomething(*t1)
}
```

We have neatly followed the convention and used value-receiver types. However, this also crashes because the caller sent in an unintialized pointer. **The conclusion is that the programmer has to be careful to avoid such pitfalls (especially in Golang). There is no *holy-cow* convention that will help avoid shooting yourself in the foot.**

Again, in this change, the scope of the struct is just one file (ignoring test files for obvious reasons). **It is quite clear that pointers will be always be initialized in this code.** So I still say that it is a fine choice to use pointer-receiver here.


> * method to pointer type is even accessible by nil pointer. Having any methods to pointer type will lead to panic.
> 
> ``` go
> func (t *mytype) greet() {
> 	fmt.Println("spock hand")
> }
> 
> // (fixed a bug here to make it compile)
> var t1 *mytype
> t1.greet()
> ```

Again, I don't understand this point - you say `"Having any methods to pointer type will lead to panic."` but just gave an example yourself of a pointer-receiver method that will [not panic or crash](https://play.golang.org/p/8NKAuImyP1). Here, it appears that it doesn't matter if the method is pointer/value-receiver, because it makes no use of any properties in the type. **Additionally, all the code in the present change are counter-examples to the statement "pointer-receiver methods will lead to panic".**

> * Choosing pointer or value type for method is by considering performance and/or changing field values.

This is not true - as I said earlier, if the method needs to change fields in the struct it has to be a pointer-receiver type. This is compulsory. There is no choice here.

The choice comes only in the case of methods not changing their type's fields. The main consideration in Go coding guidelines *is performance*! Please take some time to go through the [Go Code Review Comments](https://github.com/golang/go/wiki/CodeReviewComments#receiver-type) - they in fact say to default to pointer types if in doubt.

krishnasrinivas replied:
From [here](https://github.com/golang/go/wiki/CodeReviewComments#receiver-type):

> If the receiver is a large struct or array, a pointer receiver is more efficient. How large is large? Assume it's equivalent to passing all its elements as arguments to the method. If that feels too large, it's also too large for the receiver.

We'll need some sort of convention to decide when a struct is considered "large". We can check with AB on this.

> If the receiver is a small array or struct that is naturally a value type (for instance, something like the time.Time type), with no mutable fields and no pointers, or is just a simple basic type such as int or string, a value receiver makes sense.

This suggests that for anything other than basic types like int, string or a "naturally value type" like time.Time we should use pointer receiver.

> A value receiver can reduce the amount of garbage that can be generated; if a value is passed to a value method, an on-stack copy can be used instead of allocating on the heap. (The compiler tries to be smart about avoiding this allocation, but it can't always succeed.) Don't choose a value receiver type for this reason without profiling first.


_Don't choose a value receiver type for this reason without profiling first._

Very confusing, are they saying use pointers even for basic types like int, string and then perform profiling and then change it to value receivers if there is a performance problem?

donatello replied:
Another official resource on this topic is here - https://golang.org/doc/faq#methods_on_values_or_pointers - appears that value receivers are good (i.e. efficient and clear) for small structs according to the doc.

Also, in some cases, for value receivers, a copy of the struct may end up getting allocated on the heap rather than the stack. This creates more work for the GC as items allocated on the function's call stack are freed at the end of a function call (and don't need the GC). Go determines where to put the allocation based on [Escape Analysis](https://en.wikipedia.org/wiki/Escape_analysis). I couldn't find any official docs about the implementation but there are references to it in golang issue tracker, reddit, mailing list and other places.

balamurugana commented on cmd/update-main_test.go:
```diff
@@ -119,7 +119,10 @@ func TestGetCurrentReleaseTime(t *testing.T) {
 	}
 
```
You can add unit tests for `getUserAgent()` and `getDownloadURL()` for codecov more.

donatello replied:
Done.

balamurugana commented on cmd/update-main.go:
```diff
@@ -183,12 +203,10 @@ func downloadReleaseData(releaseChecksumURL string, timeout time.Duration, mode
-	return downloadReleaseData(minioReleaseURL+"minio.shasum", timeout, mode)
-}
+// parseReleaseData - helper to parse time of release.
+func parseReleaseData(data string) (releaseTime time.Time,
```
You could group methods of a type together for better readability.

donatello replied:
Done.

krishnasrinivas commented on cmd/update-main.go:
```diff
@@ -235,20 +256,33 @@ func getDownloadURL() (downloadURL string) {
+		cgroupFile:         "/proc/self/cgroup",
+		releaseChecksumURL: minioReleaseURL + "minio.shasum",
+		serverMode:         mode,
+	}
```
You can notice that all the fields of `updateCheckParams` (except mode) are globally available and hence there is no reason for these fields to be part of this struct.

The intention of code cleanup is correct, the previous code was looking bad i.e mode was being passed all around and in the end to be used by a function somewhere down deep. This can be fixed in a better way by having `globalMinioMode` that gets set during init. Global variables are not a problem as long as no one modifies it, so having mode as globally available is fine. Also Bala is working putting all globals into a single struct so that it looks cleaner.

So you can do away with `updateCheckParams` and have mode globally available.

donatello replied:
Yes, other than `mode` all fields are constants. If you see the previous code, this is clear there too (so why did the earlier code become complex anyway?!) - the reason for putting constant-like values as argument(s) to various functions is to be able to write tests. If these values are constants or are hard-coded then it is not possible to pass different values for the purposes of tests. See for example the test for `isDockerEnv()` - it is quite difficult to write it if the file path string "constant" `/proc/self/cgroup` is hard-coded - but clearly the test is useful and important here and explains why the function is implemented in the way it is (no need to even open `/proc/self/cgroup`!).

The problem with global variables (even those that do not change even during the lifetime of the process) is also related to writing tests. Global values impact multiple tests at the same time and so the test author needs to clearly understand which global variables are used or are affected by a function/code that is to be tested - the test author needs to ensure that before the test they are appropriately set and after the test cleaned up so it doesnt affect subsequent tests. This is generally a pain. This was discussed quite a lot in BLR office some months back with @balamurugana and @krisis - the problem with globals and why they need to be eliminated.

IMO, this is a clean implementation - serving the purpose of maintainability and testability.

krishnasrinivas replied:
There are two options:

1) Passing variables as function parameters so that the functions are unit testable. Downside is longer function signature (or use structs like updateCheckParams and have methods on it)

2) Use global variables, harder to unit test but upside is simpler function signature.

We use 2nd style in minio all over, let's discuss what style we should follow so that we can be consistent.

donatello replied:
@harshavardhana @deekoder Requesting to mark this PR as blocked on @abperiasamy

krishnasrinivas replied:
An issue related to this discussion: https://github.com/minio/minio/issues/3198

donatello committed:
Refactor update check code

- Add release-time conversion helpers

- Split GetCurrentReleaseTime() into two simpler functions.

- Avoid appending strings when assembling user-agent string.

- Reorder release info URLs to check the newer URLs earlier.

- Remove trivial low-level functions created solely for the purpose of
  writing tests.

- Remove some unnecessary tests.
